### DISCUSSION TITLE
Serverless in-browser UI via pyodide

### DISCUSSION THREAD

**Discussion by yhcharles** [zauberzeug/nicegui#2488] [Ideas / Feature Requests]
With Pyodide, one can run Python code in browser -- without backend server. This enables the possibility to create pure frontend UI with Python.

A few other Python UI frameworks has already enabled this feature:
- gradio: https://www.gradio.app/guides/gradio-lite
- panel: https://panel.holoviz.org/how_to/wasm/index.html
- streamlit: https://github.com/whitphx/stlite

How about bringing this feature to NiceGUI?

**Comment by rodja** [comment]
Wow. I was not aware of Pyodide. Looks very interesting but we currently do not have the resources to investigate a possible integration. We are very grateful for any help in this regard. A first interesting step would probably be a working prototype of Starlette or FastAPI running in the browser.

**Reply by yhcharles** [reply]
Thanks for the reply. I think we may not need to bring the whole FastAPI in browser. That could be too heavy.

I'm not familiar with the implementation details of NiceGUI, but to my understanding the UI part is done in html/JS, which communicates with backend Python server to handle user actions. If that's true, the Python server won't be necessary if using Pyodide. JS and Python can call each other directly in the browser.

**Reply by rodja** [reply]
But then you would need to rewrite all the routing code. Surely possible but would require additional maintainace.

**Reply by yhcharles** [reply]
Yes, it could be a large change. Not sure how other libraries achieved this, like gradio & panel. Maybe there's something we can learn from them.

**Comment by trungleduc** [comment]
Hi folks, I'm building a library that can serve Starlette apps in the browser, and it supports NiceGUI. Here is the discussion https://github.com/zauberzeug/nicegui/discussions/5701

**Reply by yhcharles** [reply]
Thanks! This looks very interesting and promising!

**Comment by evnchn** [comment]
> Crazy! And a bit scary. 🤯

https://github.com/zauberzeug/nicegui/pull/5776

**Comment by evnchn** [comment]
@yhcharles — this exists now 🧪 (alpha): **https://github.com/evnchn/nicegui-pyodide**

An external extension to stock NiceGUI (no fork, no core changes) that runs NiceGUI entirely in the browser via Pyodide/PyScript — **no server**. `pip install nicegui` + the extension, then `nicegui-pyodide-build ./dist` gives you a static folder to host anywhere.

**Alpha:** pinned to NiceGUI 3.14.x; verified surface is small (basic widgets, click→handler→DOM, `ui.notify`, markdown, input round-trip); no-server-only features (HTTP endpoints, static routes, native mode) don't apply. Feedback/issues welcome. It grew out of my closed core PR https://github.com/zauberzeug/nicegui/pull/5776, re-done as an external package.
